pennylane/data/__init__.py

```diff
@@ -24,7 +23,177 @@
+
+To open a dataset from a file, use :meth:`Dataset.open()` class method:
+
+    >>> my_dataset = Dataset.open("~/datasets/my_dataset.h5", mode="r")
+
+The `mode` argument follow the standard library convention - 'r' for reading, 'w-' and `w` for create and overwrite,
+and 'a' for editing. ``open()`` can be used to create a new dataset directly on disk:
+
+    >>> new_dataset = Dataset.open("~/datasets/new_datasets.h5", mode="w")
+
+By default, any changes made to an opened dataset will be committed directly to the file, which will fail
+if the file is opened read-only. The `"copy"` mode can be used to load the dataset into memory and detach
+it from the file:
+
+    >>> my_dataset = Dataset.open("~/dataset/my_dataset/h5", mode="copy")
+    >>> my_dataset.new_attribute = "abc"
+
+
+Attribute Metadata
+~~~~~~~~~~~~~~~~~~
+
+Dataset attributes can also contain additional metadata, such as docstrings. The :func:`qml.data.attribute`
+function can be used to attach metadata on assignment or initialization.
+
+    >>> hamiltonian = qml.Hamiltonian([1., 1.], [qml.PauliZ(wires=0), qml.PauliZ(wires=1)])
+    >>> eigvals, eigvecs = np.linalg.eigh(qml.matrix(hamiltonian))
+    >>> dataset = qml.data.Dataset(hamiltonian = qml.data.attribute(
+            hamiltonian,
+            doc="The hamiltonian of the system"))
+    >>> dataset.eigen = attribute(
+            {"eigvals": eigvals, "eigvecs": eigvecs},
+            doc="Eigenvalues and eigenvectors of the hamiltonain")
+
+This metadata can then be accessed using the :meth:`Dataset.attr_info` mapping:
+
+    >>> dataset.attr_info["hamiltonian"]["doc"]
+    'The hamiltonian of the system'
+
+
+Declarative API
+~~~~~~~~~~~~~~~
+
+When creating datasets to model a physical system, it is common to collect the same data for
+a system under different conditions or assumptions. For example, a collection of datasets describing
+a quantum oscillator, which contains the first 1000 energy levels for different masses and force constants.
+
+The datasets declarative API allows us to create subclasses of ``Dataset`` that define the required attributes,
+or 'fields', and their associated type and documentation:
+
+.. code-block
+
+    class QuantumOscillator(qml.data.Dataset, data_name="quantum_oscillator", params=["mass", "force_constant"]):
+        \"""Dataset describing a quantum oscillator.\"""
+
+        mass: float = qml.data.field(doc = "The mass of the particle")
+        force_constant: float = qml.data.field(doc = "The force constant of the oscillator")
+        hamiltonian: qml.Hamiltonian = qml.data.field(doc = "The hamiltonian of the particle")
+        energy_levels: np.ndarray = qml.data.field(doc = "The first 1000 energy levels of the system")
+
+The ``data_name`` keyword specifies a category or descriptive name for the dataset type, and the ``params``
+keyword to the class is used to specify fields that function as parameters, i.e they determine the behaviour
+of the system.
```

Comment:
I haven't heard about this `params` thing - is there an intention for a distinction between `params` and `attrs`? cc @DSGuala

Reply:
I think what he means by `params` is the keywords that you provide in your `qml.load` to identify your dataset uniquely. I guess `identifiers` might be a better name since stuff like `params`/`attrs` is already overloaded.

Reply:
but they behave differently, like in the oscillator example given here

Reply:
For the purposes of the datasets service that we're building, we'd like to have the parameters available on the dataset itself. So we took the view that parameters are a subset of the attributes of a dataset

Reply:
the distinction wasn't especially clear to me from the docs, but that might just be me. I guess my question is, should users every really provide a `params` value, and if so, why? I don't see any explicit use for it rn (I see implicit use/setting when doing something like `load("qchem", molname="H2")` which puts `molname` as a param)